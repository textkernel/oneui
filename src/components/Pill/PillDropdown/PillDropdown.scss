@import '../../../themes/oneui/oneui';

:root {
    --pill-dropdown-padding: var(--space-100);
}

@keyframes dropdownAppear {
    from {
        opacity: 0;
        transform: translateY(-15px) scale(0.9);
    }
    to {
        opacity: 1;
        transform: translateY(0) scale(1);
    }
}

.PillDropdown {
    z-index: 500;

    &__dialog {
        background: var(--color-background);
        margin: var(--space-75) 0;
        border-radius: var(--border-radius);
        border: 1px solid var(--color-neutral-30);
        box-shadow: 0 5px 10px 0 rgb(0 0 0 / 20%);
        transform-origin: top left;
        animation: dropdownAppear var(--transition-duration-quick) forwards;
        animation-timing-function: linear;
    }

    &__content {
        padding: var(--pill-dropdown-padding);

        &--noPadding {
            padding: 0;
        }
    }

    &__footer {
        display: flex;
        border-top: solid 1px var(--color-neutral-30);
        justify-content: flex-end;
        padding: var(--pill-dropdown-padding);
    }
}
