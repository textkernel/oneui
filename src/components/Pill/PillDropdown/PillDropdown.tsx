import * as React from 'react';
import { bem } from '../../../utils';
import { Button } from '../../Buttons';
import styles from './PillDropdown.scss';

export interface Props {
    /** The dropdown content renderer function. It is called with:
     *   * close {function} - see below
     */
    children: (param) => undefined;
    /** function that closes the dropdown */
    close: () => void;
    /** label for the Done button */
    doneLabel: string;
    /** whether or not to add a padding to the dropdown container.
     * It is useful if you need to add elements that stretch to the edge of the container.
     * You can use the CSS variable --pill-dropdown-padding to add padding as required.
     * */
    noPadding?: boolean;
}

const { block, elem } = bem('PillDropdown', styles);

export const PillDropdown = React.forwardRef<HTMLElement, Props>(
    ({ close = null, doneLabel, noPadding = false, children, ...rest }, ref) => {
        const handleDoneClick = () => {
            if (close) {
                close();
            }
        };

        return (
            <div ref={ref} role="presentation" {...rest} {...block({ noPadding, ...rest })}>
                <div role="dialog" {...elem('dialog')}>
                    <div {...elem('content', { noPadding })}>{children({ close })}</div>
                    <div {...elem('footer')}>
                        <Button context="primary" size="small" onClick={handleDoneClick}>
                            {doneLabel}
                        </Button>
                    </div>
                </div>
            </div>
        );
    }
);

PillDropdown.displayName = 'PillDropdown';
