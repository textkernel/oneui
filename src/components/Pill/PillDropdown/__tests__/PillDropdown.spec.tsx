import React from 'react';
import toJson from 'enzyme-to-json';
import { PillDropdown } from '../PillDropdown';

describe('<PillDropdown> component', () => {
    const childrenMock = jest.fn();
    const closeMock = jest.fn();

    let wrapper;

    beforeEach(() => {
        wrapper = mount(
            <PillDropdown close={closeMock} doneLabel="Done">
                {childrenMock}
            </PillDropdown>
        );
    });

    afterEach(() => {
        jest.resetAllMocks();
    });

    it('should render correctly', () => {
        expect(toJson(wrapper)).toMatchSnapshot();
    });
    it('should render correctly without padding', () => {
        wrapper.setProps({ noPadding: true });
        wrapper.update();

        expect(wrapper.find('.PillDropdown__content--noPadding')).toHaveLength(1);
    });
    it('should call children function with close as arguments', () => {
        expect(childrenMock).toHaveBeenCalledWith({
            close: closeMock,
        });
    });
});
