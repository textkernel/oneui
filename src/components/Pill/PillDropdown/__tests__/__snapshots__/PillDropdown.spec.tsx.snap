// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<PillDropdown> component should render correctly 1`] = `
<div>
  <div
    class="PillDropdown"
    role="presentation"
  >
    <div
      class="PillDropdown__dialog"
      role="dialog"
    >
      <div
        class="PillDropdown__content"
        role="group"
      />
      <div
        class="PillDropdown__footer"
      >
        <button
          class="Button Button--isPrimary Button--size_small"
          type="button"
        >
          Done
        </button>
      </div>
    </div>
  </div>
</div>
`;
